pkg/nf90io/nf90io_utils.F

```diff
@@ -0,0 +1,376 @@
+
+      CHARACTER (LEN=*), intent(in) :: fname
+      INTEGER  :: ncid, err
+
+      IF (usingMPI) THEN
+#ifdef ALLOW_USE_MPI
+         err = nf90_open(fname, IOR(nf90_write, nf90_mpiio),
+     $        ncid,
+     $        comm=MPI_COMM_WORLD, info=MPI_INFO_NULL)
+#endif
+      ELSE
+         err = nf90_open(fname, nf90_write,ncid)
+      ENDIF
+
+      NF90IO_OPEN = err
+      END FUNCTION
+
+
+
+CCCCCCCCCCCCCCC
+      SUBROUTINE NF90IO_FILL_3Dnorec(ncid, vname, dat, myThid)
+
+C     !DESCRIPTION:
+C     Convenience function for handling all MNC and NetCDF library
+C     errors.
+      use netcdf
+      implicit none
+#include "SIZE.h"
+#include "EEPARAMS.h"
+#include "PARAMS.h"
+#include "GRID.h"
+#include "NF90IO.h"
+
+
+      character*(*) vname
+      INTEGER ncid, myThid, iRec
+      _RL dat (1-OLx:sNx+OLx,1-OLy:sNy+OLy,Nr,nSx,nSy)
+
+      INTEGER start(3), count(3), rec_dimid
+      INTEGER bi, bj, err, varid
+
+      err = nf90_inq_varid(ncid, vname, varid)
+      CALL nf90ERR(err, "Getting varid",myThid)
+      err = nf90io_var_par_access(ncid, varid)
+      CALL nf90ERR(err, "Setting varid paraccess 3D",myThid)
+
+      count = (/  sNx, sNy, Nr /)
+      do bj = 1,nSy
+         do bi = 1,nSx
+C     NOT tested w/ more than one thread be processor, so
+C     maybe incorrect.
+            start = (/ myXGlobalLo+(bi-1)*sNx,
+     &           myYGlobalLo+(bj-1)*sNy,
+     &           1/)
+            err = nf90_put_var(ncid, varid,
+     &             dat(1:sNx, 1:sNy, 1:Nr, bi, bj),
+     &              start = start, count = count)
+            CALL nf90ERR(err, "Putting data into file", myThid)
+         enddo
+      enddo
+      RETURN
+      END
+
+CCCCCCCCCCCCCCCCCCCCCCCCCCCc
+
+CCCCCCCCCCCCCCC
+      SUBROUTINE NF90IO_FILL_4D(ncid, vname, dat, iRec, myThid)
+
+C     !DESCRIPTION:
+C     Convenience function for handling all MNC and NetCDF library
+C     errors.
+      use netcdf
+      implicit none
+#include "SIZE.h"
+#include "EEPARAMS.h"
+#include "PARAMS.h"
+#include "GRID.h"
+#include "NF90IO.h"
+
+
+      character*(*) vname
+      INTEGER ncid, myThid, iRec
+      _RL dat (1-OLx:sNx+OLx,1-OLy:sNy+OLy,Nr,nSx,nSy)
+
+      INTEGER start(4), count(4), rec_dimid
+      INTEGER bi, bj, err, varid
+
+      err = nf90_inq_varid(ncid, vname, varid)
+      CALL nf90ERR(err, "Getting varid",myThid)
+      err = nf90io_var_par_access(ncid, varid)
+      CALL nf90ERR(err, "Setting varid paraccess 4d",myThid)
+
+      count = (/  sNx, sNy, Nr, 1 /)
+      do bj = 1,nSy
+         do bi = 1,nSx
+C     NOT tested w/ more than one thread be processor, so
+C     maybe incorrect.
+            start = (/ myXGlobalLo+(bi-1)*sNx,
+     &           myYGlobalLo+(bj-1)*sNy,
+     &           1,
+     &           iRec /)
+            err = nf90_put_var(ncid, varid,
+     &             dat(1:sNx, 1:sNy, 1:Nr, bi, bj),
+     &              start = start, count = count)
+            CALL nf90ERR(err, "Putting data into file", myThid)
```

Comment:
My guess would be that in collective mode each processor has to make exactly one call to nf90_put_var for each record.  If that's true, one would have to combine the tiles one processor controls into one array that can be passed to nf90_put_var.  I couldn't verify this in the netCDF docs, though.  Maybe you know where to look?

Reply:
Oh, I have no idea - not claiming to be a netcdf expert, just seemed easy to put this together so I did.  

I think we could proceed for now by putting a `STOP` in somewhere.

If what you say above is correct, I'd need guidance on how frequently muti-threading is used by folks as to whether it was worth the extra code to gather each processor's output and save it.  Certainly thats substantially more complicated and memory intensive than what is done now.  If its worth it (i.e. *many* of people actually multithread) then we should do it as a future refinement.  Myself, I'm not even sure *how* to run a multithreaded job, though no doubt I could figure it out; maybe as easy as just setting `nSx/nSy`?

Reply:
This is not just for multi-threading, one can run with multiple tiles per processor in a single thread (that's what these loops are for).  Sometimes it is convenient if one can run on fewer processors without having to change the tiling.  And testreport uses this to run parallel tests on as many processors as there are available.

Reply:
Confirmed on my machine.  Advice appreciated.  We could gather a processer-global temp variable and `nf90_put_var` that variable instead.  I think that seems reasonable, and not too much work.

Reply:
I think I found my bug and/or implimented gathering.  Not sure gathering is necessary.  Basically, I needed to trim the arrays `qdiag` to `Nr` because they are `NrMax` in the k-direction, and that caused bad data to be in the routines if `sNx>1`...

Reply:
Yeah, didn't need to gather, just needed to pass the properly trimmed data to the writing functions. @jahn, this should work now.  

Working on making the example `testreport` ready.

Reply:
yes it does!